scratch-js/renderer/Shaders.mjs

```diff
@@ -0,0 +1,229 @@
+const SpriteShader = {};
+
+SpriteShader.vertex = `
+precision mediump float;
+
+attribute vec2 a_position;
+uniform mat3 u_transform;
+uniform vec2 u_stageSize;
+
+varying vec2 v_texCoord;
+
+void main() {
+  v_texCoord = vec2(a_position.x, 1.0 - a_position.y);
+  gl_Position = vec4((u_transform * vec3(a_position, 1.0)) / vec3(u_stageSize * 0.5, 1.0), 1.0);
+}
+`;
+
+SpriteShader.fragment = `
+precision mediump float;
+
+const float epsilon = 1e-3;
+
+uniform sampler2D u_texture;
+varying vec2 v_texCoord;
+
+#ifdef EFFECT_color
+uniform float u_color;
+#endif
+
+#ifdef EFFECT_fisheye
+uniform float u_fisheye;
+#endif
+
+#ifdef EFFECT_whirl
+uniform float u_whirl;
+#endif
+
+#ifdef EFFECT_pixelate
+uniform float u_pixelate;
+uniform vec2 u_skinSize;
+#endif
+
+#ifdef EFFECT_mosaic
+uniform float u_mosaic;
+#endif
+
+#ifdef EFFECT_brightness
+uniform float u_brightness;
+#endif
+
+#ifdef EFFECT_ghost
+uniform float u_ghost;
+#endif
+
+#if defined(EFFECT_whirl) || defined(EFFECT_fisheye)
+const vec2 CENTER = vec2(0.5, 0.5);
+#endif
+
+#ifdef DRAW_MODE_COLOR_MASK
+uniform vec4 u_colorMask;
+
+// TODO: Scratch 2.0 and Scratch 3.0's CPU path check if the top 6 bits match,
+// which a tolerance of 3/255 should be equivalent to,
+// but Scratch's GPU path has a tolerance of 2/255.
```

Comment:
That's actually super interesting! This PR is teaching me a bunch of cool stuff. 😛